gtsam/sfm/tests/testShonanAveraging.cpp

```diff
@@ -0,0 +1,367 @@
+  EXPECT_DOUBLES_EQUAL(0, kShonan.cost(SO3Values), 1e-4);
+}
+
+/* ************************************************************************* */
+TEST(ShonanAveraging3, MakeATangentVector) {
+  Vector9 v;
+  v << 1, 2, 3, 4, 5, 6, 7, 8, 9;
+  Matrix expected(5, 5);
+  expected << 0, 0, 0, 0, -4, //
+      0, 0, 0, 0, -5,         //
+      0, 0, 0, 0, -6,         //
+      0, 0, 0, 0, 0,          //
+      4, 5, 6, 0, 0;
+  const Vector xi_1 = ShonanAveraging3::MakeATangentVector(5, v, 1);
+  const auto actual = SOn::Hat(xi_1);
+  CHECK(assert_equal(expected, actual));
+}
+
+/* ************************************************************************* */
+TEST(ShonanAveraging3, LiftTo) {
+  auto I = genericValue(Rot3());
+  Values initial{{0, I}, {1, I}, {2, I}};
+  Values lifted = ShonanAveraging3::LiftTo<Rot3>(5, initial);
+  EXPECT(assert_equal(SOn(5), lifted.at<SOn>(0)));
+}
+
+/* ************************************************************************* */
+TEST(ShonanAveraging3, LiftwithDescent) {
+  ShonanAveraging3::Parameters parameters;
+  // remove gauge so eigenvalue well defined:
+  parameters.setAnchorWeight(1000);
+  parameters.setKarcherWeight(1000);
+  parameters.setGaugesWeight(1000);
+  string g2oFile = findExampleDataFile("Klaus3.g2o");
+  auto shonan = ShonanAveraging3(g2oFile, parameters);
+
+  // control randomness
+  static std::mt19937 rng(0);
+  const Values randomRotations = shonan.initializeRandomly(rng);
+  Values random = ShonanAveraging3::LiftTo<Rot3>(3, randomRotations);
+
+  // Optimize
+  const Values Qstar3 = shonan.tryOptimizingAt(3, random);
+
+  // Compute Eigenvalue with Spectra solver
+  double lambda = shonan.computeMinEigenValue(Qstar3);
+
+  // Check Eigenvalue with slow Eigen version, converts matrix A to dense matrix
+  // :-(
+  bool computeEigenvectors = true;
+  const Matrix S = ShonanAveraging3::StiefelElementMatrix(Qstar3);
+  auto A = shonan.computeA(S);
+  Eigen::EigenSolver<Matrix> eigenSolver(Matrix(A), computeEigenvectors);
+  auto lambdas = eigenSolver.eigenvalues().real();
+  int index = 0;
+  double minEigenValue = lambdas(0);
+  for (int i = 1; i < lambdas.size(); i++)
+    if (lambdas(i) < minEigenValue) {
+      minEigenValue = lambdas(i);
+      index = i;
+    }
+  EXPECT_DOUBLES_EQUAL(minEigenValue, lambda, 1e-12);
+
+  // Get predictable eigenvector from Eigen
+  Vector minEigenVector = eigenSolver.eigenvectors().col(index).real();
+
+  // Check descent with eigen version
+  Values initialQ4 =
+      ShonanAveraging3::LiftwithDescent(4, Qstar3, minEigenVector);
+  EXPECT_LONGS_EQUAL(3, initialQ4.size());
+  Matrix expected(4, 4);
+  expected << 0.88891, 0.0257057, -0.0639449, -0.452868,  //
+      0.0257057, 0.994052, 0.0147965, 0.104791,           //
+      -0.0639455, 0.0147967, 0.963192, -0.260678,         //
+      0.452868, -0.104791, 0.260678, 0.846154;
+  EXPECT(assert_equal(SOn(expected), initialQ4.at<SOn>(0), 1e-5));
+}
```

Comment:
@david-m-rosen here is the descent check that fails

Reply:
OK got it. Let me put some thought behind this then -- it should be possible to fix this up without requiring that we use any particular eigenvector ...

Reply:
OK, if this case is meant to be a unit test **just** for the LiftwithDescent method, then it can be radically simplified.  As currently implemented, LiftwithDescent just performs the lift of the input indicated in eq. (27) of the paper, constructs the tangent vector from eq. (28), and then performs retraction -- that is, this function doesn't actually care whether the input v is a minimum eigenvector or not.  

I would suggest just replacing this with a simple hard-coded example involving an initial set of 2 or 3 rotations, a corresponding fixed vector v, and then just verifying that the returned result is what it should be.  This would entirely avoid the issue of having to compute **specific** eigenvectors across multiple platforms

Reply:
That sounds like a plan :-)